### PR TITLE
feat: cross-snapshot Extensions dependencies

### LINKED ISSUE #17768: core: When snapshotting, the list of already included extensions is lost
https://github.com/denoland/deno/blob/9e3d433249b9259e3a04b4f68563a41455ac7efc/cli/build.rs#L355-L359

This makes it impossible to specify dependent extensions between snapshots.

### PR DESCRIPTION
**Pull request by crowlKats**
Closes #17768

### CODE REVIEW

File: cli/build.rs [hunk 1]
```diff
@@ -366,7 +366,7 @@ fn create_cli_snapshot(snapshot_path: PathBuf) {
     // FIXME(bartlomieju): information about which extensions were
     // already snapshotted is not preserved in the snapshot. This should be
     // fixed, so we can reliably depend on that information.
-    // .dependencies(vec!["runtime"])
+    .dependencies(vec!["deno_websocket"])
```
Comment (bartlomieju) [inline cli/build.rs:369]: Why changed to `deno_websocket` instead of runtime?

File: core/runtime.rs [hunk 2]
```diff
@@ -454,10 +460,17 @@ impl JsRuntime {
 
       let mut module_handles = vec![];
       let mut scope = v8::ContextScope::new(scope, context);
-      // The 0th element is the module map itself, followed by X number of module
+
+      let extensions: Vec<String> =
```
Comment (bartlomieju) [inline core/runtime.rs:464]: Add comment here

File: core/runtime.rs [hunk 3]
```diff
@@ -432,13 +433,18 @@ impl JsRuntime {
...
 
     fn get_context_data(
       scope: &mut v8::HandleScope<()>,
       context: v8::Local<v8::Context>,
-    ) -> (Vec<v8::Global<v8::Module>>, v8::Global<v8::Object>) {
+    ) -> (
+      Vec<String>,
+      Vec<v8::Global<v8::Module>>,
+      v8::Global<v8::Object>,
+    ) {
```
Comment (bartlomieju) [inline core/runtime.rs:447]: Feels like we should move this whole function to `snapshot_util`?

**Review by bartlomieju — COMMENTED** [review]
Can you update one of the snapshot tests to verify this functionality?

File: cli/build.rs (hunk 1 above)
Comment (crowlKats) [reply cli/build.rs:369]: runtime isnt an ext

Comment (bartlomieju) [reply cli/build.rs:369]: It seems it is: https://github.com/denoland/deno/blob/0aeb8bc7591ec32e199b98f4532d6a962c6da6cb/runtime/build.rs#L169

Comment (crowlKats) [reply cli/build.rs:369]: oh you are right; overlooked as its defined in the snapshot. will revert it to runtime then

**Comment by bartlomieju** [comment]
After talking with Divy we're not sure we should land this - this will cause to add a hefty amount of V8 strings that will need to be deserialized on each startup.

**Comment by aapoalas** [comment]
> After talking with Divy we're not sure we should land this - this will cause to add a hefty amount of V8 strings that will need to be deserialized on each startup.

You recently changed the extension names to use static external string resources, right? Would it be possible to "pull those back" to avoid string deserialisation and access the static string pointer directly?

**Comment by aapoalas** [comment]
> That's a good question, we should investigate that. I think we might be able to use `v8::String::write_one_byte`

With the C++ API it's possible to do one better:
1. Get the external one byte string resource from a String: https://v8docs.nodesource.com/node-18.2/d2/db3/classv8_1_1_string.html#a790ceedd8a57ed7b12b1c2c2e6775faf
2. Get the data pointer and length of the one byte string contained within the resource: https://v8docs.nodesource.com/node-18.2/d9/db3/classv8_1_1_string_1_1_external_one_byte_string_resource.html#ad21f4b84b5c0731695992e6701defaa1
3. Use the string as you see fit without a single copy.

The Rust API does not currently offer the necessary APIs to get a resource from a String but it would not be hard to implement those. For future reference it would be a good idea to somehow implement a C++/Rust bridging Resource base / trait that could be used by both Deno and Embedders.

One mistake in the C++ API that Node unfortunately exposes is that there is no way to tell apart different subclasses of the OneByteStringResource. Thus, if an embedder uses this subclass and extends its interface in their own way (say, they link it with a reference counted string struct), then they must keep track of each of the pointers of these instances they've created. Otherwise they cannot safely "downcast" lest they accidentally operate on an instance that was not created by them but by the runtime itself.

**Comment by bartlomieju** [comment]
After using new snapshot system and `extension!` macro for the past few weeks I think this PR is not needed after all. I'm gonna close without a merge.

**Pull request closed without merging**